### ISSUE TITLE
About splitting heads for temporal and spatial modeling

### ISSUE THREAD

**Issue by kaiw7** [leexinhao/ZeroI2V#3]
Hi, could I learn about whether your implementation is based official openai CLIP script? Because I noticed that they implement the multi-head attention with inner pytorch function so that the modificaton of multi-head attention operation is a little complex. Many thanks.

**Comment by leexinhao** [comment]
Yes, we directly modified the `nn.MultiHeadAttention` in PyTorch.

**Comment by kaiw7** [comment]
> Yes, we directly modified the `nn.MultiHeadAttention` in PyTorch.

Thank you very much for your quick reply. Could I know which part you modified in nn.MultiHeadAttention function in PyTorch?

**Comment by leexinhao** [comment]
Mainly using shift operations in the `forward`, we will release the source code in a while.

**Comment by kaiw7** [comment]
> Mainly using shift operations in the `forward`, we will release the source code in a while.

Thanks a lot. Do you mean the 'F.multi_head_attention_forward' function?

**Comment by kaiw7** [comment]
And also, when you directly modified the 'nn.MultiHeadAttention' in PyTorch, do you first make a copy of torch library or directly add/modify something in corresponding inner function of pytorch using IDE?

**Comment by leexinhao** [comment]
> And also, when you directly modified the 'nn.MultiHeadAttention' in PyTorch, do you first make a copy of torch library or directly add/modify something in corresponding inner function of pytorch using IDE?

No, we just copied a python implementation of the `nn.MultiHeadAttention` and modified it.

**Comment by kaiw7** [comment]
> > And also, when you directly modified the 'nn.MultiHeadAttention' in PyTorch, do you first make a copy of torch library or directly add/modify something in corresponding inner function of pytorch using IDE?
> 
> No, we just copied a python implementation of the `nn.MultiHeadAttention` and modified it.

Many thanks, do you also copy these two functions in the `nn.MultiHeadAttention` as below?
from .module import Module
from .. import functional as F

**Comment by Yusics** [comment]
Just a following up question related to the shift implementation.
Are you using torch.roll for the shift operation? Thanks!
